From 331b0f77b40864fa3553c56ec6cee50c65afeedb Mon Sep 17 00:00:00 2001
From: Monika Korada <quic_koramoni@quicinc.com>
Date: Sat, 3 Feb 2024 12:50:20 +0530
Subject: [PATCH] [WAR] hostapd: fix drv corruption due to choosing invalid
 wiphy_data

After insmod, the pointers of older wiphy_data pointing to drv are
expected to be removed. But here the pointers of the older wiphy_data
are still pointing to drv which is resulting in drv corruption.

This fix will remove the links pointing to drv from the older wiphy_data.

Signed-off-by: Monika Korada <quic_koramoni@quicinc.com>
---
 src/drivers/driver_nl80211.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -960,9 +960,11 @@ add:
 			break;
 		}
 	}
-	/* if not add it */
-	if (!found)
+	/* if not add it by deleting the older wiphy */
+	if (!found) {
+		dl_list_del(&bss->drv->wiphy_list);
 		dl_list_add(&w->drvs, &bss->drv->wiphy_list);
+	}
 
 	dl_list_add(&w->bsss, &bss->wiphy_list);
 	bss->wiphy_data = w;
@@ -2264,6 +2266,7 @@ static void * wpa_driver_nl80211_drv_ini
 	drv->ctx = ctx;
 	drv->hostapd = !!hostapd;
 	drv->eapol_sock = -1;
+	dl_list_init(&drv->wiphy_list);
 
 	/*
 	 * There is no driver capability flag for this, so assume it is
